src/main/java/com/p6spy/engine/spy/appender/CustomLineFormat.java

```diff
@@ -51,27 +97,106 @@
-      return FALLBACK_FORMATTING_STRATEGY.formatMessage(connectionId, now, elapsed, category, prepared, sql, url);
+      return FALLBACK_FORMATTING_STRATEGY.formatMessage(connectionId, now, elapsed, category, prepared, sql, url, attributes);
+    }
+
+    // If the log format has changed, make sure that only one thread parses the new style.
+    // We can safely compare pointers to the format String, because it is cached in the active instance until it changes.
+    if (lastProcessedFormat != customLogMessageFormat && formatParsingInProgress.compareAndSet(false, true)) {
+      try {
+        parseNewFormat(customLogMessageFormat);
+      } finally {
+        lastProcessedFormat = customLogMessageFormat;
+        formatParsingInProgress.set(false);
+      }
     }
 
-    return customLogMessageFormat
-      .replaceAll(Pattern.quote(CONNECTION_ID), Integer.toString(connectionId))
-      .replaceAll(Pattern.quote(CURRENT_TIME), now)
-      .replaceAll(Pattern.quote(EXECUTION_TIME), Long.toString(elapsed))
-      .replaceAll(Pattern.quote(CATEGORY), category)
-      .replaceAll(Pattern.quote(EFFECTIVE_SQL), Matcher.quoteReplacement(prepared))
-      .replaceAll(Pattern.quote(EFFECTIVE_SQL_SINGLELINE), Matcher.quoteReplacement(P6Util.singleLine(prepared)))
-      .replaceAll(Pattern.quote(SQL), Matcher.quoteReplacement(sql))
-      .replaceAll(Pattern.quote(SQL_SINGLE_LINE), Matcher.quoteReplacement(P6Util.singleLine(sql)))
-      .replaceAll(Pattern.quote(URL), url);
+    Map<String, String> replacements = new HashMap<String, String>() {{
+      put(CONNECTION_ID, Integer.toString(connectionId));
+      put(CURRENT_TIME, now);
+      put(EXECUTION_TIME, Long.toString(elapsed));
+      put(CATEGORY, category);
+      put(EFFECTIVE_SQL, Matcher.quoteReplacement(prepared));
+      put(SQL, Matcher.quoteReplacement(sql));
+      put(URL, url);
+
+      if (isUsed_EFFECTIVE_SQL_SINGLELINE) {
+        put(EFFECTIVE_SQL_SINGLELINE, Matcher.quoteReplacement(P6Util.singleLine(prepared)));
+      }
+
+      if (isUsed_SQL_SINGLE_LINE) {
+        put(SQL_SINGLE_LINE, Matcher.quoteReplacement(P6Util.singleLine(sql)));
+      }
+
+      if (isUsed_ATTRIBUTES && ! usedAttributesPlaceholders.isEmpty()) {
+        // We're only expanding the first placeholder for the Attributes and the rest will be printed empty
+        put(usedAttributesPlaceholders.get(0), attributes != null ? attributes.toString() : "");
+      }
+    }};
+
+    StringBuffer output = new StringBuffer();
+    try {
+      Matcher m = AllPlaceholdersPattern.matcher(lastProcessedFormat);
+
+      while (m.find()) {
+        String replacement = replacements.get(m.group());
+        m.appendReplacement(output, replacement != null ? replacement : "");
+      }
+      m.appendTail(output);
+    } catch (Exception ex) {
+      // Make sure that any exceptions from the Matcher are not breaking the main application
+System.out.println("=== Got Exception : " + Arrays.toString(ex.getStackTrace()));
```

Comment:
All "System.out.println" calls are going to be removed.  Forgot to clean up while learning how to make GitHub forks and pull requests.